ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,381 @@
+                        vnodeFsidInode,
+                        cacheIndex,    // starting index
+                        indexFromHash, // stopping index
+                        vnodeVid,
+                        invalidateEntry,
+                        /* out */ rootHandle))
+                {
+                    ClearCacheAndInsertEntry_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        vnode,
+                        vnodeFsidInode,
+                        indexFromHash,
+                        vnodeVid,
+                        /* out */ rootHandle);
+                }
+            }
+        }
+        else
+        {
+            perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+            
+            UpgradeToExclusiveLock(s_entriesLock);
+            lockElevatedToExclusive = true;
+            
+            ClearCacheAndInsertEntry_ExclusiveLocked(
+                perfTracer,
+                cacheMissFallbackFunctionCounter,
+                cacheMissFallbackFunctionInnerLoopCounter,
+                vnode,
+                vnodeFsidInode,
+                indexFromHash,
+                vnodeVid,
+                /* out */ rootHandle);
+        }
+    }
+    
+    if (lockElevatedToExclusive)
+    {
+        RWLock_ReleaseExclusive(s_entriesLock);
+    }
+    else
+    {
+        RWLock_ReleaseShared(s_entriesLock);
+    }
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nullable perfTracer)
+{
+    if (perfTracer)
+    {
+        perfTracer->IncrementCount(PrjFSPerfCounter_CacheInvalidateCount, true /*ignoreSampling*/);
+    }
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        InvalidateCache_ExclusiveLocked();
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
+{
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+}
+
+static inline void UpgradeToExclusiveLock(RWLock& lock)
+{
+    if (!RWLock_AcquireSharedToExclusive(lock))
+    {
+        RWLock_AcquireExclusive(lock);
+    }
+}
+
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode)
+{
+    uintptr_t vnodeAddress = reinterpret_cast<uintptr_t>(vnode);
+    return (vnodeAddress >> 3) % s_entriesCapacity;
+}
+
+KEXT_STATIC bool TryFindVnodeIndex_SharedLocked(
+    vnode_t _Nonnull vnode,
+    uintptr_t startingIndex,
+    uintptr_t stoppingIndex,
+    /* out parameters */
+    uintptr_t& cacheIndex)
+{
+    // Walk from the starting index until we find:
+    //    -> The vnode
+    //    -> A NULLVP entry
+    //    -> The stopping index
+    // If we hit the end of the array, continue searching from the start
+    cacheIndex = startingIndex;
+    while (vnode != s_entries[cacheIndex].vnode)
+    {
+        if (NULLVP == s_entries[cacheIndex].vnode)
```

Comment:
Why return true if we hit NULLVP?  Doesn't this mean we didn't find it?

Reply:
It took me a while to work this one out too - the return value essentially indicates whether `cacheIndex` has returned a valid index, not whether the vnode is already stored at that bucket. The idea seems to be that we don't need to locate the bucket twice for the search-then-insert case (i.e. an optimisation).

I think if we fail to upgrade the lock we'll pretty much have to start over anyway, so perhaps this can be simplified in the case where we do manage to upgrade the lock, because no repeated search is needed for insertion, we know no other thread will have beaten us to it.

Reply:
pmj is correct.  The idea is that we're trying to find the index for the vnode, and there are two scenarios:

- The vnode is already in the cache
- The vnode is not yet in the cache (in which case the index is where the vnode should be inserted)

@jeschu1 let me know if you think the current changes are any easier to follow.  We have to handle both of the above scenarios because at the point we're looking up the index we don't know if the vnode is already in the cache (even if we failed to look it up another thread might have come along and inserted it between us dropping the shared lock and acquiring the exclusive lock).

Reply:
Marking this comment as resolved, @jeschu1, @pmj please re-open if you'd like to see additional changes in this function.